libsigner/src/events.rs

```diff
@@ -0,0 +1,268 @@
+// Copyright (C) 2013-2020 Blockstack PBC, a public benefit corporation
+// Copyright (C) 2020-2023 Stacks Open Internet Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::sync::mpsc::Sender;
+
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+
+use std::net::SocketAddr;
+use std::net::TcpListener;
+use std::net::TcpStream;
+
+use std::io::Read;
+
+use clarity::vm::types::QualifiedContractIdentifier;
+use libstackerdb::StackerDBChunkData;
+
+use serde::{Deserialize, Serialize};
+
+use crate::http::{decode_http_body, decode_http_request};
+use crate::EventError;
+
+/// Event structure for newly-arrived StackerDB data
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct StackerDBChunksEvent {
+    pub contract_id: QualifiedContractIdentifier,
+    pub modified_slots: Vec<StackerDBChunkData>,
+}
+
+/// Trait to implement a stop-signaler for the event receiver thread.
+/// The caller calls `send()` and the event receiver loop (which lives in a separate thread) will
+/// terminate.
+pub trait EventStopSignaler {
+    fn send(&mut self);
+}
+
+/// Trait to implement to handle StackerDB events sent by the Stacks node
+pub trait EventReceiver {
+    /// The implementation of ST will ensure that a call to ST::send() will cause
+    /// the call to `is_stopped()` below to return true.
+    type ST: EventStopSignaler + Send + Sync;
+
+    /// Open a server socket to the given socket address.
+    fn bind(&mut self, listener: SocketAddr) -> Result<SocketAddr, EventError>;
+    /// Return the next event
+    fn next_event(&mut self) -> Result<StackerDBChunksEvent, EventError>;
+    /// Add a downstream event consumer
+    fn add_consumer(&mut self, event_out: Sender<StackerDBChunksEvent>);
+    /// Forward the event to downstream consumers
+    fn forward_event(&mut self, ev: StackerDBChunksEvent) -> bool;
```

Comment:
What's the intention of the returned bool? It seems to me like we'd want to have a result returned here to convey more information as I assume forwarding an event is a fallible operation.

Reply:
Sure, personally I'd prefer to convey that information more explicitly as
```
fn forward_event(&mut self, ev: ...) -> Result<(), ForwardError>;

enum ForwardError{
  ConsumerChannelBroken
}
```

Although this is more verbose, it's clearer to read and conveys more information about the system. Regarding complexity, the `Result<(), ForwardError>` has cardinality of two so it is as simple as a boolean.